packages/hash/frontend/src/blocks/page/createPlaceholderPlugin/createPlaceholderPlugin.tsx

```diff
@@ -0,0 +1,96 @@
+import { paragraphBlockComponentId } from "@hashintel/hash-shared/blocks";
+import {
+  componentNodeToId,
+  findComponentNodes,
+} from "@hashintel/hash-shared/prosemirror";
+import { Schema } from "prosemirror-model";
+import { Plugin, PluginKey } from "prosemirror-state";
+import { Decoration, DecorationSet } from "prosemirror-view";
+import { RenderPortal } from "../usePortals";
+import { Placeholder } from "./Placeholder";
+
+interface PlaceholderPluginState {
+  focused: boolean;
+}
+
+const placeholderPluginKey = new PluginKey<PlaceholderPluginState, Schema>(
+  "placeholderPlugin",
+);
+
+export const createPlaceholderPlugin = (renderPortal: RenderPortal) => {
+  return new Plugin<PlaceholderPluginState, Schema>({
+    key: placeholderPluginKey,
+    state: {
+      init() {
+        return { focused: false };
+      },
+      apply(tr, state) {
+        const focused = tr.getMeta(placeholderPluginKey);
+
+        // return the old state if transaction does not have focused meta
+        if (focused === undefined) {
+          return state;
+        }
+
+        return { focused };
+      },
+    },
+    props: {
+      decorations(state) {
+        const firstNode = state.selection.$anchor.node(1);
+
+        if (!firstNode) return;
+
+        const componentId = componentNodeToId(
+          findComponentNodes(firstNode)[0]!,
+        );
+
+        const isFocused = placeholderPluginKey.getState(state)?.focused;
+        const isParagraph = componentId === paragraphBlockComponentId;
+        const isEmpty = findComponentNodes(firstNode)[0]?.childCount === 0;
```

Comment:
`findComponentNodes(firstNode)[0]` is duplicated here, but also in one place we assert it definitely exists (`!`) and in another we use optional property access. 

I'm also a bit wary of `paragraphBlockComponentId` spreading around the codebase. I think I'd prefer a function which returns true/false.

Reply:
Ooops I missed the duplicate, thanks! About `paragraphBlockComponentId `, do you suggest a function like `isParagraphNode(node)` ?

Reply:
Updated here https://github.com/hashintel/hash/pull/953/commits/5bebff617f57205a74e1f4ccba39a2e062624377